### PR TITLE
Skia: Improve rendering quality of layers and qt style with fractional translations

### PR DESCRIPTION
**Pull request by tronical**
Layers and cached pixmaps should be "invisible" in terms of quality,
and they're "designed" to be pixel aligned.

### CODE REVIEW

**Review by ogoffart — APPROVED** [review]
:+1:

But i'm wondering if we should't keep the error from accumulating.
Could we keep the fractional part somehow in a state? or keep the translation accurate, but draw the image slightly shifted?
But this is probably still an improvement.

**Comment by tronical** [comment]
Yes, good point. Let me try a different way :)

**Comment by tronical** [comment]
Slightly more invasive approach. At least for subsequent translations I think that could work. But OTOH this still has the issue that the children of an item will be aligned relative to the aligned parent.

Not much of a fan of this approach either, but I see the issue with accumulation.

File: internal/renderers/skia/itemrenderer.rs [hunk 1]
```diff
@@ -1006,3 +1004,126 @@ fn adjust_rect_and_border_for_inner_drawing(
...
+
+        let Some(device_to_local) = local_to_device.invert() else {
+            return;
+        };
+        let mut target_point = local_to_device.map_point(skia_safe::Point::default());
+
+        target_point.x = target_point.x.round();
+        target_point.y = target_point.y.round();
+
+        self.actual_renderer.canvas.translate(device_to_local.map_point(target_point));
```
Comment (ogoffart) [inline internal/renderers/skia/itemrenderer.rs:1027]: So this rely on the fact that we save and restore the state between every item?

**Review by ogoffart — COMMENTED** [review]
Indeed this is a bit intrusive to have a wrapper around the whole renderer.
The alternative would be to define pixel_align_origin in the SkiaItemRenderer, add a call to `self.pixel_align_origin()` in every function that needs drawing.
For a pure line line of code count, it'd still be smaller because you anyway need at least one line for each wrapper now.
But it might be a bit more error prone.

Some call such as visit_clip, visit_opacity, may actually not need it.  Perhaps draw text neither since the glyphs will anyway be drawn with float coordinates (although maybe it's better for alignment)

**Review by ogoffart — APPROVED** [review]
In principle, we should still restore before drawging the children.
But layer shouldn't have children, and qt widget probably don't either. But it could look like something like that:
```
let _something_that_restore_in_drop = self.pixel_align_origin();
```
(and pixel_align_origin saves)

**Comment by tronical** [comment]
Oh you’re right. Thanks! I’ll fix that up.

**Review by ogoffart — APPROVED** [review]

**Pull request merged**